src/app/InteractionModelEngine.cpp

```diff
@@ -76,21 +78,17 @@ void InteractionModelEngine::Shutdown()
 
     mCommandHandlerList = nullptr;
 
-    //
-    // Since modifying the pool during iteration is generally frowned upon,
-    // I've chosen to just destroy the object but not necessarily de-allocate it.
-    //
-    // This poses a problem when shutting down and restarting the stack, since the
-    // IMEngine is a statically constructed singleton, and this lingering state will
-    // cause issues.
-    //
-    // This doesn't pose a problem right now because there shouldn't be any actual objects
-    // left here due to the synchronous nature of command handling.
-    //
-    // Filed #10332 to track this.
-    //
-    mCommandHandlerObjs.ForEachActiveObject([](CommandHandler * obj) -> bool {
-        obj->~CommandHandler();
+    // Increase magic number to invalid all Handle-s.
+    mMagic++;
+
+    mCommandHandlerObjs.ForEachActiveObject([this](CommandHandler * obj) -> bool {
+        // Modifying the pool during iteration is generally frowned upon.
+        // This is almost safe since mCommandHandlerObjs is a BitMapObjectPool which won't malfunction when modifying the inner
+        // record while during traversal. But this behavior is not guranteed, so we should fix this by implementing DeallocateAll.
+        //
+        // TODO(@kghost, #10332) Implement DeallocateAll and replace this.
+
+        mCommandHandlerObjs.Deallocate(obj);
```

Comment:
Please add a comment that says that CommandHandler relies on Shutdown _not_ being called here?

That said, if we get here, what closes the exchange?  Followup to get that sorted out?